community-content/cpr-examples/timm_serving/build.py

```diff
@@ -0,0 +1,99 @@
+# Copyright 2022 Google LLC
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     https://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Build the model server container."""
+import os
+import pathlib
+from typing import Sequence
+
+from absl import app
+from absl import logging
+from config import CPRConfig
+from google.cloud import aiplatform
+from google.cloud.aiplatform import prediction as cpr
+import smart_open
+import timm
+from timm_serving import predictor
+import torch
+
+def build_container(config: CPRConfig, tag: str) -> cpr.LocalModel:
+  """Build the model server container.
+
+  Args:
+    tag: Output image tag.
+
+  Returns:
+    LocalModel exposing the built model server.
+  """
+  return cpr.LocalModel.create_cpr_model(
+      src_dir=os.path.join(os.getcwd()),
+      output_image_uri=tag,
+      base_image=config.base_image,
+      predictor=predictor.TimmPredictor,
+      requirements_path=os.path.join(os.getcwd(), "requirements.txt"),
+  )
+
+def save_model_artifact(destination: str) -> None:
+  """Save a copy of the model state dict."""
+  model = timm.create_model(predictor.TimmPredictor.TIMM_MODEL_NAME,
+                            pretrained=True)
+  dest_file = os.path.join(destination, 
+                           predictor.TimmPredictor.WEIGHTS_FILE)
+  with smart_open.open(dest_file, "wb") as f:
+    torch.save(model, f)  
+  logging.info("Saved model to %s", dest_file)
+  logging.info("%s parameters", sum(p.numel() for p in model.parameters()))
+
+
+def upload_model(config: CPRConfig) -> aiplatform.Model:
+  """Tag and upload the model server."""
+  ar_tag = (f"{config.region}-docker.pkg.dev/{config.project_id}"
+            f"/{config.repository}/{config.image}")
+  local_model = build_container(config, tag=ar_tag)
```

Comment:
Ah, I have a better understanding of your comment now... about the persistence of the local model when iterating in python scripts.

Yea, it's unfortunate that we have to build container again to upload model.

Reply:
or saving the container spec in config and re-initializing LocalModel without re-building images?

Reply:
This is actually not the issue I was complaining about -- rebuilding the image takes very little time if no changes were made, since the layers are all already in the docker cache.

The stateful config is there to persist the uploaded model name between executions. If we don't do that, I have to upload and deploy in the same invocation.